### ISSUE TITLE
Rancher host starts with wrong hostname

### ISSUE THREAD

**Issue by palbiez** [rancher/os#2404]
**RancherOS Version: (ros os version)** v1.4.0
**Where are you running RancherOS? (docker-machine, AWS, GCE, baremetal, etc.)** VMWare

We are working with 3 rancherOS hosts and one Ubuntu host. 
All hosts are working with DHCP and get a fixed IP by DHCP server
At every time we're restarting one of our rancherOS hosts we have a wrong entry at DNS. 
Entry in DNS should be rancher = 10.0.2.19
After reboot from rancherOS 
DNS Entry is rancher = 10.0.1.xx Number of rancherOS host. 

We changed the hostname at installation to rancherworkerX. Also we override the default hostname in boot.yml.

Can you please check and tell me if you need any logfiles?

Thanks in advance

**Comment by niusmallnan** [comment]
If you do not specify a nameserver in cloud-config, RancherOS will try to get the nameserver via DHCP.

**Comment by palbiez** [comment]
@niusmallnan I know. And nameserver via DHCP is the right one. The problem  is, that rancherOS send DHCP requests with hostname rancher and not as rancherworkerX

**Comment by niusmallnan** [comment]
> At every time we're restarting one of our rancherOS hosts we have a wrong entry at DNS.
Entry in DNS should be rancher = 10.0.2.19
After reboot from rancherOS
DNS Entry is rancher = 10.0.1.xx Number of rancherOS host.

Oh, I didn't understand this part. 

You mean that you have configured the hostname `rancherworkerX`, but the dns entry obtained after reboot is `rancher=<ip>` instead of the expected `rancherworkerX=<ip>`?
How did you get the DNS entry? 
Can you give more information about the operation steps? So I can try to reproduce it.

**Comment by palbiez** [comment]
@niusmallnan 
DNS and DHCP is provided by PFSense at 10.0.1.2
Before reboot of rancherOS Host
`nslookup rancher
Server:         10.0.1.2
Address:        10.0.1.2#53

Name:   rancher.XXXXX
Address: 10.0.2.19
`
After reboot 
```
nslookup rancher
Server:         10.0.1.2
Address:        10.0.1.2#53

Name:   rancher.XXX
Address: 10.0.1.21
```
and 
nslookup rancherworker2
Server:         10.0.1.2
Address:        10.0.1.2#53

Name:   rancherworker2.XXX
Address: 10.0.1.21

At dhcpd.conf there is a fixed entry for the rancherworkers and rancher-server
```
host s_lan_2 {
        hardware ethernet 42:e5:4c:ed:14:65;
        option dhcp-client-identifier "rancherworker2";
        fixed-address 10.0.1.21;
        option host-name "rancherworker2";

host s_opt1_11 {
        hardware ethernet 96:61:8c:89:67:8a;
        option dhcp-client-identifier "rancher";
        fixed-address 10.0.2.19;
```

Here are my cloud-config.yml
`rancher:
  ssh:
    keys:
      dsa: |+
      dsa-pub: |+
        ssh-dss 
root@rancherworker2
      ecdsa: |+
      ecdsa-pub: |+
        ecdsa-sha2-nistp256 
X+wIs6k= root@rancherworker2
root@rancherworker2
      rsa: |+
root@rancherworker2
write_files:
  - path: /etc/docker/certs.d/nexus.acando.tech/ca.crt
    permissions: "0644"
    owner: root
    content: |
...
  - container: ntp
    path: /etc/ntp.conf
    permissions: "0644"
    owner: root
    content: |
      server 10.0.1.1 iburst
      server 10.0.2.2 iburst
runcmd:
- mkdir persistentstore
- mount -o port=2049,nolock,proto=tcp 10.0.2.19:/persistentstore /persistentstore
`
I only truncated the ssh keys

```
 ros config get hostname
rancherworker2
```

**Comment by palbiez** [comment]
I rebooted rancherworker1 now. Here is the dhcpd logfile from pfs

```
Jul 18 07:45:19 pfs dhcpd: DHCPREQUEST for 10.0.1.20 from 36:4d:5c:ad:2d:d9 via vtnet1
Jul 18 07:45:19 pfs dhcpd: DHCPACK on 10.0.1.20 to 36:4d:5c:ad:2d:d9 via vtnet1
Jul 18 07:45:19 pfs dhcpd: Added new forward map from rancherworker1.acando.tech to 10.0.1.20
Jul 18 07:45:19 pfs dhcpd: Unable to add reverse map from 20.1.0.10.in-addr.arpa. to rancherworker1.acando.tech: REFUSED
Jul 18 07:47:29 pfs dhcpd: DHCPREQUEST for 10.0.2.19 from 96:61:8c:89:67:8a via vtnet2
Jul 18 07:47:29 pfs dhcpd: DHCPACK on 10.0.2.19 to 96:61:8c:89:67:8a via vtnet2
Jul 18 07:47:29 pfs dhcpd: Added new forward map from rancher.acando.tech to 10.0.2.19
Jul 18 07:48:03 pfs dhcpd: Wrote 0 deleted host decls to leases file.
Jul 18 07:48:03 pfs dhcpd: Wrote 0 new dynamic host decls to leases file.
Jul 18 07:48:03 pfs dhcpd: Wrote 42 leases to leases file.
Jul 18 07:49:21 pfs dhcpd: DHCPREQUEST for 10.0.1.21 from 42:e5:4c:ed:14:65 via vtnet1
Jul 18 07:49:21 pfs dhcpd: DHCPACK on 10.0.1.21 to 42:e5:4c:ed:14:65 via vtnet1
Jul 18 07:49:21 pfs dhcpd: Added new forward map from rancherworker2.acando.tech to 10.0.1.21
Jul 18 07:49:21 pfs dhcpd: Unable to add reverse map from 21.1.0.10.in-addr.arpa. to rancherworker2.acando.tech: REFUSED
Jul 18 07:52:53 pfs dhcpd: DHCPREQUEST for 10.0.1.22 from 92:e6:25:ca:1f:a0 via vtnet1
Jul 18 07:52:53 pfs dhcpd: DHCPACK on 10.0.1.22 to 92:e6:25:ca:1f:a0 via vtnet1
Jul 18 07:52:53 pfs dhcpd: Added new forward map from rancherworker3.acando.tech to 10.0.1.22
Jul 18 07:52:53 pfs dhcpd: Unable to add reverse map from 22.1.0.10.in-addr.arpa. to rancherworker3.acando.tech: REFUSED

Jul 18 08:17:23 pfs dhcpd: DHCPDISCOVER from 36:4d:5c:ad:2d:d9 via vtnet1
Jul 18 08:17:23 pfs dhcpd: DHCPOFFER on 10.0.1.20 to 36:4d:5c:ad:2d:d9 via vtnet1
Jul 18 08:17:23 pfs dhcpd: DHCPREQUEST for 10.0.1.20 (10.0.1.2) from 36:4d:5c:ad:2d:d9 via vtnet1
**Jul 18 08:17:23 pfs dhcpd: DHCPACK on 10.0.1.20 to 36:4d:5c:ad:2d:d9 via vtnet1
Jul 18 08:17:23 pfs dhcpd: Added new forward map from rancher.acando.tech to 10.0.1.20**
Jul 18 08:17:23 pfs dhcpd: Unable to add reverse map from 20.1.0.10.in-addr.arpa. to rancher.acando.tech: REFUSED
Jul 18 08:17:27 pfs dhcpd: DHCPREQUEST for 10.0.1.20 from 36:4d:5c:ad:2d:d9 via vtnet1
Jul 18 08:17:27 pfs dhcpd: DHCPACK on 10.0.1.20 to 36:4d:5c:ad:2d:d9 via vtnet1
Jul 18 08:17:27 pfs dhcpd: Added new forward map from rancherworker1.acando.tech to 10.0.1.20
Jul 18 08:17:27 pfs dhcpd: Unable to add reverse map from 20.1.0.10.in-addr.arpa. to rancherworker1.acando.tech: REFUSED
```
The MAC from rancherworker1 is 36:4d:5c:ad:2d:d9

resolver.log contains the following lines
```
**Jul 18 08:17:23 pfs named[84862]: update: info: client @0x804411a00 10.0.1.2#54261/key pfs.acando.tech: view internal: updating zone 'acando.tech/IN': adding an RR at 'rancher.acando.tech' A 10.0.1.20**
Jul 18 08:17:27 pfs named[84862]: update: info: client @0x802a73c00 10.0.1.2#54261/key pfs.acando.tech: view internal: updating zone 'acando.tech/IN': adding an RR at 'rancherworker1.acando.tech' A 10.0.1.20
```


If I want to connect to rancher.acando.tech I get the following message from Hypervisor

```
ssh rancher
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@       WARNING: POSSIBLE DNS SPOOFING DETECTED!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
The ECDSA host key for rancher has changed,
and the key for the corresponding IP address 10.0.1.20
is unchanged. This could either mean that
DNS SPOOFING is happening or the IP address for the host
and its host key have changed at the same time.
Offending key for IP in /root/.ssh/known_hosts:55
  remove with:
  ssh-keygen -f "/root/.ssh/known_hosts" -R 10.0.1.20
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!
Someone could be eavesdropping on you right now (man-in-the-middle attack)!
It is also possible that a host key has just been changed.
The fingerprint for the ECDSA key sent by the remote host is

Please contact your system administrator.
Add correct host key in /root/.ssh/known_hosts to get rid of this message.
Offending ECDSA key in /root/.ssh/known_hosts:68
  remove with:
  ssh-keygen -f "/root/.ssh/known_hosts" -R rancher
ECDSA host key for rancher has changed and you have requested strict checking.
Host key verification failed.
```

At best you can see at init.log on rancherOS host:

```
[rancher@rancherworker1 boot]$ less init.log | grep rancher
time="2018-07-18T06:17:17Z" level=debug msg="Config: EXTRA_CMDLINE: /init
rancher:
      image: rancher/os-bootstrap:v1.1.3
        io.rancher.os.detach: "false"
        io.rancher.os.scope: system
      - /var/lib/rancher:/var/lib/rancher:ro
      image: rancher/os-base:v1.1.3
        io.rancher.os.detach: "false"
        io.rancher.os.scope: system
      - /etc/ssl/certs/ca-certificates.crt:/etc/ssl/certs/ca-certificates.crt.rancher
      - /var/lib/rancher:/var/lib/rancher
      - /var/lib/rancher/conf:/var/lib/rancher/conf
    **hostname: rancher**
```

10 seconds later at netconf.log
```
time="2018-07-18T06:17:27Z" level=debug msg="getDhcpLease eth0: broadcast_address=10.0.1.255
dhcp_lease_time=7200
dhcp_message_type=5
dhcp_server_identifier=10.0.1.2
domain_name=acando.tech
domain_name_servers=10.0.1.2
**host_name=rancherworker1**
ip_address=10.0.1.20
network_number=10.0.1.0
routers=10.0.1.2
subnet_cidr=24
subnet_mask=255.255.255.0
"
```

**Comment by palbiez** [comment]
```
less user_config.yml | grep hostname
hostname: rancherworker1

# less /usr/share/ros/os-config.yml | grep hostname
    hostname: rancherworker

# less init.yml | grep hostname

```

**Comment by niusmallnan** [comment]
RancherOS uses the default hostname `rancher` when booting, after getting the new hostname via DHCP, RancherOS will update the hostname.

I don‘t have a PFSense setup, but there are two ways that you can change the default hostname.

1. add a kernel parameter `rancher.hostname=rancherxworker`,  
2. update the `hostname` in  ` /usr/share/ros/os-config.yml`

These allow RancherOS to use the newly configured hostname during the earliest boot process.

**Comment by palbiez** [comment]
Thanks for this workaround. 
We use a workaround in PFSense now. 
Force dynamic DNS hostname to be the same as configured hostname for Static Mappings.

**Comment by OnyxCorpINC** [comment]
I've been bashing my head for hours now to try and get the correct `hostname` picked up. The best I managed to achieve while using only `cloud-config` was rebooting and the script ran and worked. However it only happened once and its not the most elegant. 

The best workaround I found at this point is to just launch a container and mount the hosts `/etc/hostname`. Then I feed that info a variable like this for example `vmhostname=$( cat /etc/hostname )` and I launch some sort of deployment a job lets say and install whatever I need to run the script.

I am sort of giving doing things in `cloud-config` that are too exotic and thinking this method is more versatile. I can put an annotation to allow node selection and I can download any script I want and run it.

I wish I would have thought of this two weeks ago as I've been massaging `cloud-config` to work as I want but either I don't understand how to leverage it or this is all intentional so that I would figure out that I should leverage running some container to do what I need.

If there is a better solution using just `cloud-config` I'd love to find out!
